File: src/main/java/me/rsman/BetterMinecraftCore/configs/models/BmcCraftSubContainer.java
package me.rsman.BetterMinecraftCore.configs.models;

import me.rsman.BetterMinecraftCore.configs.containers.BmcCraftContainer;
import me.rsman.BetterMinecraftCore.configs.containers.BmcItemContainer;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Tag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.RecipeChoice;

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class BmcCraftSubContainer {
    private Map<String, BmcShapedCraft> shaped;
    private Map<String, BmcShapelessCraft> shapeless;

    public BmcCraftSubContainer() {}

    public Map<String, BmcShapedCraft> getShaped() {
        return shaped;
    }

    public void setShaped(Map<String, BmcShapedCraft> shaped) {
        this.shaped = shaped;
    }

    public Map<String, BmcShapelessCraft> getShapeless() {
        return shapeless;
    }

    public void setShapeless(Map<String, BmcShapelessCraft> shapeless) {
        this.shapeless = shapeless;
    }
    public static ItemStack convertItemSchemeToItemStack(String item){
        if(item.equals("m.AIR") || item.equals("AIR") || item.equals("null")){
            return null;
        } else if(item.startsWith("m.")){
            if(Material.matchMaterial(item.substring(2)) == null) return null;
            return new ItemStack(Objects.requireNonNull(Material.matchMaterial(item.substring(2))));
        } else {
            if(BmcItemContainer.getInstance().getItems().containsKey(item)){
                return BmcItemContainer.getInstance().getItems().get(item).getItemStack();
            } else {
                return null;
            }
        }
    }

    public static Map.Entry<Character, RecipeChoice> convertSchemeToRecipeChoice(String scheme, Character character) {
        Map.Entry<Character, RecipeChoice> out;
        String[] splitted = scheme.trim().split(" ");
        String item = splitted[0];
        int number = splitted.length == 1 ? 1 : Integer.parseInt(splitted[1]);
        if(item.equals("m.AIR") || item.equals("AIR") || item.equals("null")){
            out = new AbstractMap.SimpleEntry<>(' ', null);
        } else if(item.startsWith("m.")){
            if(Material.matchMaterial(item.substring(2)) == null) return null;
            out = new AbstractMap.SimpleEntry<>(character, new RecipeChoice.ExactChoice(new ItemStack(Objects.requireNonNull(Material.matchMaterial(item.substring(2))), number)));
        }  else if(item.startsWith("all.")){
            boolean added = false;
            Tag<Material> tag =  Bukkit.getTag("blocks", NamespacedKey.minecraft(item.substring(4).toLowerCase()), Material.class);
            assert tag != null;
            if(tag.getValues().toArray().length == 0){
                tag =  Bukkit.getTag("items", NamespacedKey.minecraft(item.substring(4).toLowerCase()), Material.class);
            }
            assert tag != null;
            if(tag.getValues().toArray().length == 0){
                tag =  Bukkit.getTag("fluids", NamespacedKey.minecraft(item.substring(4).toLowerCase()), Material.class);
            }
            assert tag != null;
            out = new AbstractMap.SimpleEntry<>(character, new RecipeChoice.MaterialChoice(tag));
        } else {
            if(BmcItemContainer.getInstance().getItems().containsKey(item)){
                ItemStack temp = BmcItemContainer.getInstance().getItems().get(item).getItemStack();
                temp.setAmount(number);
                out = new AbstractMap.SimpleEntry<>(character, new RecipeChoice.ExactChoice(temp));
            } else {
                return null;
            }
        }
        return out;
    }

    public BmcCraftSubContainer cloneForConfig() {
        BmcCraftSubContainer csc = new BmcCraftSubContainer();
        if (shaped != null) {
            HashMap<String, BmcShapedCraft> shapedClone = new HashMap<>();
            for (Map.Entry<String, BmcShapedCraft> shapedToClone: shaped.entrySet()) {
                shapedClone.put(shapedToClone.getKey(), shapedToClone.getValue().cloneForConfig());
            }
            csc.setShaped(shapedClone);
        }
        if (shapeless != null) {
            HashMap<String, BmcShapelessCraft> shapelessClone = new HashMap<>();
            for (Map.Entry<String, BmcShapelessCraft> shapelessToClone: shapeless.entrySet()) {
                shapelessClone.put(shapelessToClone.getKey(), shapelessToClone.getValue().cloneForConfig());
            }
            csc.setShapeless(shapelessClone);
        }
        return csc;
    }
}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/configs/models/BmcCraftSubContainer.kt
package me.rsman.BetterMinecraftCore.configs.models

import me.rsman.BetterMinecraftCore.BetterMinecraftCore.Companion.instance
import me.rsman.BetterMinecraftCore.enums.EAttributes
import me.rsman.BetterMinecraftCore.enums.EEnchants
import me.rsman.BetterMinecraftCore.configs.models.BmcItem
import org.bukkit.inventory.ItemStack
import org.bukkit.Material
import me.rsman.BetterMinecraftCore.Managers.ItemManager
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.RecipeChoice
import me.rsman.BetterMinecraftCore.configs.models.BmcCraftSubContainer
import me.rsman.BetterMinecraftCore.BetterMinecraftCore
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.inventory.RecipeChoice.ExactChoice
import org.bukkit.NamespacedKey
import org.bukkit.inventory.RecipeChoice.MaterialChoice
import me.rsman.BetterMinecraftCore.configs.models.BmcShapedCraft
import me.rsman.BetterMinecraftCore.configs.containers.BmcCraftContainer
import org.bukkit.inventory.ShapelessRecipe
import me.rsman.BetterMinecraftCore.configs.models.BmcShapelessCraft
import me.rsman.BetterMinecraftCore.configs.containers.BmcItemContainer
import java.util.AbstractMap.SimpleEntry
import org.bukkit.Bukkit
import me.rsman.BetterMinecraftCore.configs.ConfigLoader
import me.rsman.BetterMinecraftCore.configs.containers.GlobalConfigContainer
import me.rsman.BetterMinecraftCore.configs.containers.EnchantLangContainer
import me.rsman.BetterMinecraftCore.configs.containers.AttributeLangContainer
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.CustomClassLoaderConstructor
import java.io.FileInputStream
import org.yaml.snakeyaml.DumperOptions
import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.io.FileOutputStream
import org.yaml.snakeyaml.representer.Representer
import org.yaml.snakeyaml.nodes.NodeTuple
import java.util.*

class BmcCraftSubContainer {
    var shaped: MutableMap<String?, BmcShapedCraft?>? = null
    var shapeless: MutableMap<String?, BmcShapelessCraft?>? = null
    fun cloneForConfig(): BmcCraftSubContainer {
        val csc = BmcCraftSubContainer()
        if (shaped != null) {
            val shapedClone = HashMap<String?, BmcShapedCraft?>()
            for ((key, value) in shaped!!) {
                shapedClone[key] = value!!.cloneForConfig()
            }
            csc.shaped = shapedClone
        }
        if (shapeless != null) {
            val shapelessClone = HashMap<String?, BmcShapelessCraft?>()
            for ((key, value) in shapeless!!) {
                shapelessClone[key] = value!!.cloneForConfig()
            }
            csc.shapeless = shapelessClone
        }
        return csc
    }

    companion object {
        @JvmStatic
        fun convertItemSchemeToItemStack(item: String): ItemStack? {
            return if (item == "m.AIR" || item == "AIR" || item == "null") {
                null
            } else if (item.startsWith("m.")) {
                if (Material.matchMaterial(item.substring(2)) == null) null else ItemStack(Objects.requireNonNull(Material.matchMaterial(item.substring(2)))!!)
            } else {
                if (BmcItemContainer.instance?.items?.containsKey(item) == true) {
                    BmcItemContainer.instance!!.items?.get(item)?.itemStack
                } else {
                    null
                }
            }
        }

        fun convertSchemeToRecipeChoice(scheme: String?, character: Char): Map.Entry<Char, RecipeChoice?>? {
            val out: Map.Entry<Char, RecipeChoice?>
            val splitted = scheme!!.trim { it <= ' ' }.split(" ".toRegex()).toTypedArray()
            val item = splitted[0]
            val number = if (splitted.size == 1) 1 else splitted[1].toInt()
            if (item == "m.AIR" || item == "AIR" || item == "null") {
                out = SimpleEntry<Char, RecipeChoice?>(' ', null)
            } else if (item.startsWith("m.")) {
                if (Material.matchMaterial(item.substring(2)) == null) return null
                out = SimpleEntry<Char, RecipeChoice?>(character, ExactChoice(ItemStack(Objects.requireNonNull(Material.matchMaterial(item.substring(2)))!!, number)))
            } else if (item.startsWith("all.")) {
                var tag = Bukkit.getTag("blocks", NamespacedKey.minecraft(item.substring(4).toLowerCase()), Material::class.java)?:return null
                if (tag.values.toTypedArray().size == 0) {
                    tag = Bukkit.getTag("items", NamespacedKey.minecraft(item.substring(4).toLowerCase()), Material::class.java)?:return null
                }
                if (tag.values.toTypedArray().size == 0) {
                    tag = Bukkit.getTag("fluids", NamespacedKey.minecraft(item.substring(4).toLowerCase()), Material::class.java)?:return null
                }
                out = SimpleEntry<Char, RecipeChoice?>(character, MaterialChoice(tag))
            } else {
                if (BmcItemContainer.instance?.items?.containsKey(item) == true) {
                    val temp: ItemStack = BmcItemContainer.instance?.items?.get(item)?.itemStack ?: return null
                    temp.amount = number
                    out = SimpleEntry<Char, RecipeChoice?>(character, ExactChoice(temp))
                } else {
                    return null
                }
            }
            return out
        }
    }
}